common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts

```diff
@@ -39,26 +46,76 @@ export default class LDMLKeyboardXMLSourceFileReader {
       }
     }
     boxXmlArray(source?.keyboard?.reorders, 'reorder');
-    this.boxImportsAndSpecials(source);
+    this.boxImportsAndSpecials(source, 'keyboard');
     return source;
   }
 
   /**
    * Recurse over object, boxing up any specials or imports
    * @param obj any object to be traversed
+   * @param subtag the leafmost enclosing tag such as 'keyboard'
    */
-  private boxImportsAndSpecials(obj: any) {
+  private boxImportsAndSpecials(obj: any, subtag: string) {
     if (!obj) return;
     if (Array.isArray(obj)) {
       for (const sub of obj) {
-        this.boxImportsAndSpecials(sub);
+        // retain the same subtag
+        this.boxImportsAndSpecials(sub, subtag);
       }
     } else if(typeof obj === 'object') {
       for (const key of Object.keys(obj)) {
-        if (key === 'special' || key === 'import') {
+        if (key === 'special') {
+          boxXmlArray(obj, key);
+        } else if(key === 'import') {
+          // Need to 'box it up' first for processing
           boxXmlArray(obj, key);
+          // Now, resolve the import
+          this.resolveImports(obj, subtag);
+          // now delete the import array we so carefully constructed, the caller does not
+          // want to see it.
+          delete obj['import'];
         } else {
-          this.boxImportsAndSpecials(obj[key]);
+          this.boxImportsAndSpecials(obj[key], key);
+        }
+      }
+    }
+  }
+
+  private resolveImports(obj: any, subtag: string) {
+    for (const anImport /* : LKImport */ of obj['import']) {
+      const asImport: LKImport = anImport;
+      if (asImport.base !== 'cldr') {
+        throw new Error(`import element with base ${asImport.base} is unsupported.`);
+      }
+      const paths = asImport.path.split('/');
+      if (!paths[0] || !paths[1] || paths.length !== 2) {
```

Comment:
Testing zero-length strings? Might be clearer to do:

```suggestion
      if (paths[0] == '' || paths[1] == '' || paths.length !== 2) {
```

Reply:
thanks